o/snapstate: add ~/Snap init related to backend

MiguelPires opened the pull request:
Adds backend functions related to ~/Snap initialization and removal.

mardy commented on dirs/dirs.go:
```diff
@@ -145,12 +145,16 @@ const (
 	// HiddenSnapDataHomeDir is an experimental hidden directory for snap data
 	HiddenSnapDataHomeDir = ".snap/data"
 
+	// ExposedSnapDir is the directory where snaps should place user-facing
+	// data after ~/snap has been migrated to ~/.snap
```
nitpick: I guess it's obvious, but it might be better to specify that this is relative to the user's `$HOME`.

mardy commented on overlord/snapstate/backend/copydata.go:
```diff
@@ -239,3 +239,91 @@ var removeIfEmpty = func(dir string) error {
+			return err
+		}
+
+		if maybeFailForTesting != nil {
+			osutil.MustBeTestBinary("'maybeFailForTesting' should only be used in tests")
+			if err := maybeFailForTesting(); err != nil {
+				return err
+			}
+		}
```
mmm... I can't say that I like this a lot :-)

Since you are using a couple of methods below which would trigger a similar behaviour (`osutil.MkdirAllChown` and `ioutil.ReadDir`), why not mock them instead?

mardy commented on snap/info.go:
```diff
@@ -228,6 +228,11 @@ func UserSnapDir(home string, name string, opts *dirs.SnapDirOptions) string {
 }
 
+// UserExposedHomeDir returns the snap's directory in the exposed home dir.
+func UserExposedHomeDir(home string, name string) string {
```
nitpick: I'd rename the `name` variable to `snapName`, or given that all other methods use the same convention, keep it as it is bu then change the docstring to explain what `name` can be (the functions above all say "The name can be either a snap name or snap instance name." — but I'm not sure whether this applies here too).

mardy reviewed (commented):
Added a few minor comments inline, but I miss the big picture here. At first sight it looks like we are getting rid of `~/snap/` just to introduce `~/Snap/`, which I hope is not the case :-)

MiguelPires commented:
Thanks for the feedback @mardy, I've addressed your comments.

> Added a few minor comments inline, but I miss the big picture here. At first sight it looks like we are getting rid of `~/snap/` just to introduce `~/Snap/`, which I hope is not the case :-)

My bad, I should've added more context to the PR description. This PR was split out of https://github.com/snapcore/snapd/pull/11383 and the broader plan is described in https://forum.snapcraft.io/t/experimental-flag-for-hiding-snap/28509. The current ~/snap will be split into ~/.snap/data and ~/Snap, which will serve different purposes.

stolowski commented on dirs/dirs.go:
```diff
@@ -165,8 +169,13 @@ var (
 	HiddenSnapDataDir bool
+
+	// MigratedToExposedHome determines if the snap's directory in ~/Snap has been
+	// initialized.
```
Did you mean "migrated" here?

stolowski replied:
From the code below it looks like "initialized" is deliberate here, maybe expand the comment a bit about what "initialized" means.

MiguelPires replied:
No, I thought "initialized" might be clearer in this case. For this part of the migration, it's not a directory moving from one place to another, ~/Snap is being created and initialized with the contents from the revisioned directory.

MiguelPires replied:
Done

stolowski commented on overlord/snapstate/backend/copydata.go:
```diff
@@ -239,3 +242,82 @@ var removeIfEmpty = func(dir string) error {
+		return err
+	}
+
+	for _, usr := range users {
+		uid, gid, err := osutil.UidGid(usr)
+		if err != nil {
+			return err
+		}
+
+		newUserHome := snap.UserExposedHomeDir(usr.HomeDir, snapName)
+		if err := mkdirAllChown(newUserHome, 0700, uid, gid); err != nil {
+			return fmt.Errorf("cannot create %q: %v", newUserHome, err)
+		}
+
+		userData := snap.UserDataDir(usr.HomeDir, snapName, rev, opts)
+		entries, err := ioutil.ReadDir(userData)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				// there's nothing to copy into ~/Snap/<snap> (like on a fresh install)
+				continue
+			}
+
+			return err
+		}
+		for _, e := range entries {
+			src := filepath.Join(userData, e.Name())
+			dst := filepath.Join(newUserHome, e.Name())
+
+			if err := osutil.CopyFile(src, dst, osutil.CopyFlagPreserveAll|osutil.CopyFlagSync); err != nil {
+				return err
+			}
+		}
+	}
```
Should we revert any of that if we fail on one of the users?

MiguelPires replied:
Yes but that has to be done as part of the migration logic in the handlers (using the RemoveExposedSnapHome function)

MiguelPires replied:
It can be seen in handlers.go in the draft PR https://github.com/snapcore/snapd/pull/11383

stolowski commented on overlord/snapstate/backend/copydata.go:
```diff
@@ -239,3 +242,82 @@ var removeIfEmpty = func(dir string) error {
+			src := filepath.Join(userData, e.Name())
+			dst := filepath.Join(newUserHome, e.Name())
+
+			if err := osutil.CopyFile(src, dst, osutil.CopyFlagPreserveAll|osutil.CopyFlagSync); err != nil {
```
Interestingly, our CopyFile seems to work with directories only with CopyFlagPreserveAll flag because then we use `cp -a` internally. Without the flag CopyFile would expect a file and fail.

MiguelPires replied:
Thanks for this. I've changed the test to also have to copy a directory, to make sure that it fails if that flag is removed.

stolowski commented on overlord/snapstate/backend/copydata.go:
```diff
@@ -239,3 +242,82 @@ var removeIfEmpty = func(dir string) error {
+		exposedSnapDir := filepath.Dir(newUserHome)
+		if err := removeIfEmpty(exposedSnapDir); err != nil {
+			handle(fmt.Errorf("cannot remove %q: %v", exposedSnapDir, err))
+			continue
```
Nitpick, this `continue` can be dropped.

MiguelPires replied:
True, it's just to avoid someone forgetting to add it when adding more code after this block

stolowski commented on overlord/snapstate/backend_test.go:
```diff
@@ -1249,6 +1249,16 @@ func (f *fakeSnappyBackend) UndoHideSnapData(snapName string) error {
 	return f.maybeErrForLastOp()
 }
 
+func (f *fakeSnappyBackend) InitExposedSnapHome(snapName string, rev snap.Revision) error {
+	f.appendOp(&fakeOp{op: "init-exposed-snap-home", name: snapName})
+	return f.maybeErrForLastOp()
+}
+
+func (f *fakeSnappyBackend) RemoveExposedSnapHome(snapName string) error {
+	f.appendOp(&fakeOp{op: "rm-exposed-snap-home", name: snapName})
+	return f.maybeErrForLastOp()
```
Nice

stolowski reviewed (approved):
Looks good.

MiguelPires commented:
@mardy when you have time, can you have another look at this please?

mardy reviewed (approved):
Sorry, my review inbox got a bit too large :-) LGTM!

pedronis commented on dirs/dirs.go:
```diff
@@ -145,12 +145,16 @@ const (
 
+	// ExposedSnapDir is the directory where snaps should place user-facing
+	// data after ~/snap has been migrated to ~/.snap
+	ExposedSnapDir = "Snap"
```
shouldn't this name contain Home?

MiguelPires replied:
Renamed to ExposedSnapHomeDir

pedronis reviewed (commented):
name question
